### DISCUSSION TITLE
How to configure consumer push

### DISCUSSION THREAD

**Discussion by Vitor-Henrique-EM** [nats-io/nats.net#471] [General]
In the old version of the Nats.Client library, there was a way to use pull and push consumers.
When we had a stream context in which we wanted to create a push consumer, we used it as follows:

"jetStream.PushSubscribeAsync(subject, handler)".

In this new version, how do push consumers work?
I can't specify a subject for the consumer to listen to?

For example:

StreamConfig optionsStream = new(nameStream, ["DOMAIN.CHANNEL.*", "DOMAIN.RESPONSIBLE.*"]);
INatsJSStream jetStream = await jetStreamContext.CreateStreamAsync(optionsStream);

INatsJSConsumer consumer = await jetStream.CreateOrUpdateConsumerAsync(new ConsumerConfig()
{
    DeliverPolicy = ConsumerConfigDeliverPolicy.New
});

await foreach (NatsJSMsg<string> msg in consumer.ConsumeAsync<string>()) 
//Here at this point the consumer will receive all messages published on the channel, and there is no way to specify a specific subject

**Comment by mtmk** [comment]
we try to hide the consumer internals, that's why we do not expose any concept of pull or push consumers. If you want us to expose that please feel free to create an issue so we can look into it.

**Comment by geekox86** [comment]
@mtmk Hi. Are the latest consumers considered to be Pull or Push ones?
Out of curiosity, why did you decide to hide these concepts and their options?

**Reply by mtmk** [reply]
all consumers are pull based. there is a general consensus that pull consumers give more control over flow and back pressure handling. however we have internal push consumer implementations where we have to use for kv watchers for example. we can consider opening them as low level APIs or options if there is interest.

**Reply by Jarema** [reply]
If we expose them here, we would need to expose them in every other client in the new API. Let's discuss this first.

**Reply by geekox86** [reply]
I can see in the API docs under JetStream/Model there are config classes that cover pretty much everything in the NATS docs including Push/Pull consumers. Is this considered internal or old API?

BTW and maybe slightly off-topic but I'd appreciate a quick explanation, I still cannot get my head around why push consumers do not load balance by default when more than one client subscribe to them just like the straight-forward pull consumers? I mean why do we need DeliverSubject and DeliverGroup?
